lib/gssapi/mech/gss_add_cred.c

```diff
@@ -93,94 +93,132 @@ gss_add_cred(OM_uint32 *minor_status,
 	    *acceptor_time_rec = 0;
 	if (actual_mechs)
 	    *actual_mechs = GSS_C_NO_OID_SET;
 
-	new_cred = malloc(sizeof(struct _gss_cred));
-	if (!new_cred) {
-		*minor_status = ENOMEM;
-		return (GSS_S_FAILURE);
-	}
-	HEIM_SLIST_INIT(&new_cred->gc_mc);
-
-	/*
-	 * We go through all the mc attached to the input_cred_handle
-	 * and check the mechanism. If it matches, we call
-	 * gss_add_cred for that mechanism, otherwise we copy the mc
-	 * to new_cred.
-	 */
-	target_mc = 0;
-	if (cred) {
-		HEIM_SLIST_FOREACH(mc, &cred->gc_mc, gmc_link) {
-			if (gss_oid_equal(mc->gmc_mech_oid, desired_mech)) {
-				target_mc = mc;
-			}
-			copy_mc = _gss_copy_cred(mc);
-			if (!copy_mc) {
-				release_cred = (gss_cred_id_t)new_cred;
-				gss_release_cred(&junk, &release_cred);
-				*minor_status = ENOMEM;
-				return (GSS_S_FAILURE);
-			}
-			HEIM_SLIST_INSERT_HEAD(&new_cred->gc_mc, copy_mc, gmc_link);
-		}
-	}
-
-	/*
-	 * Figure out a suitable mn, if any.
-	 */
-	if (desired_name) {
-		major_status = _gss_find_mn(minor_status,
-					    (struct _gss_name *) desired_name,
-					    desired_mech,
-					    &mn);
-		if (major_status != GSS_S_COMPLETE) {
-			free(new_cred);
-			return major_status;
-		}
-	} else {
-		mn = 0;
-	}
-
-	m = __gss_get_mechanism(desired_mech);
-
-	mc = malloc(sizeof(struct _gss_mechanism_cred));
-	if (!mc) {
-		release_cred = (gss_cred_id_t)new_cred;
-		gss_release_cred(&junk, &release_cred);
-		*minor_status = ENOMEM;
-		return (GSS_S_FAILURE);
-	}
-	mc->gmc_mech = m;
-	mc->gmc_mech_oid = &m->gm_mech_oid;
-
-	major_status = m->gm_add_cred(minor_status,
-	    target_mc ? target_mc->gmc_cred : GSS_C_NO_CREDENTIAL,
-	    desired_name ? mn->gmn_name : GSS_C_NO_NAME,
-	    desired_mech,
-	    cred_usage,
-	    initiator_time_req,
-	    acceptor_time_req,
-	    &mc->gmc_cred,
-	    actual_mechs,
-	    initiator_time_rec,
-	    acceptor_time_rec);
-
-	if (major_status) {
-		_gss_mg_error(m, major_status, *minor_status);
-		release_cred = (gss_cred_id_t)new_cred;
-		gss_release_cred(&junk, &release_cred);
-		free(mc);
-		return (major_status);
-	}
-	HEIM_SLIST_INSERT_HEAD(&new_cred->gc_mc, mc, gmc_link);
-	*output_cred_handle = (gss_cred_id_t) new_cred;
-
-	return (GSS_S_COMPLETE);
+        if (input_cred_handle == GSS_C_NO_CREDENTIAL &&
+            output_cred_handle == NULL) {
+            return GSS_S_CALL_INACCESSIBLE_WRITE;
+        }
+
+        if ((m = __gss_get_mechanism(desired_mech)) == NULL)
+            return GSS_S_BAD_MECH;
+
+        /* NO */
+        if (output_cred_handle == NULL) {
+            HEIM_SLIST_FOREACH(mc, &cred->gc_mc, gmc_link) {
+                if (gss_oid_equal(mc->gmc_mech_oid, desired_mech))
+                    return GSS_S_COMPLETE;
+            }
+            return GSS_S_COMPLETE;
+        }
+
+
+        if (output_cred_handle) {
+            new_cred = malloc(sizeof(struct _gss_cred));
+            if (!new_cred) {
+                *minor_status = ENOMEM;
+                return (GSS_S_FAILURE);
```

Comment:
return GSS_S_FAILURE

Reply:
picky I know

Reply:
Oh, I agree.

Reply:
Pushed a separate commit to restyle this file.